maro/rl/training/async_tools/actor.py

```diff
@@ -0,0 +1,147 @@
+    if isinstance(agent_wrapper, AgentWrapper):
+        msg = SessionMessage(MsgTag.GET_INITIAL_POLICY_STATE, proxy.name, policy_server_address)
+        reply = proxy.send(msg)[0]
+        policy_version = reply.body[MsgKey.VERSION]
+        agent_wrapper.set_policy_states(reply.body[MsgKey.POLICY_STATE])
+
+    # main loop
+    for ep in range(1, num_episodes + 1):
+        t0 = time.time()
+        num_experiences_collected = 0
+        agent_wrapper.explore()
+        env_wrapper.reset()
+        env_wrapper.start()  # get initial state
+        segment = 0
+        while env_wrapper.state:
+            segment += 1
+            logger.info(
+                f"Collecting simulation data (episode {ep}, segment {segment}, policy version {policy_version})"
+            )
+            start_step_index = env_wrapper.step_index + 1
+            steps_to_go = num_steps
+            while env_wrapper.state and steps_to_go:
+                env_wrapper.step(agent_wrapper.choose_action(env_wrapper.state))
+                steps_to_go -= 1
+
+            logger.info(
+                f"Roll-out finished (episode {ep}, segment {segment}, "
+                f"steps {start_step_index} - {env_wrapper.step_index})"
+            )
+
+            exp_by_agent = env_wrapper.get_experiences()
+            policies_with_new_exp = agent_wrapper.on_experiences(exp_by_agent)
+            num_experiences_collected += sum(exp.size for exp in exp_by_agent.values())
+            exp_by_policy = agent_wrapper.get_experiences_by_policy(policies_with_new_exp)
+            reply = proxy.send(
+                SessionMessage(
+                    MsgTag.COLLECT_DONE, proxy.name, policy_server_address,
+                    body={MsgKey.EXPERIENCES: exp_by_policy, MsgKey.VERSION: policy_version}
+                )
+            )[0]
+            policy_version = reply.body[MsgKey.VERSION]
+            agent_wrapper.set_policy_states(reply.body[MsgKey.POLICY_STATE])
+
+        # update the exploration parameters
+        agent_wrapper.exploration_step()
+
+        # performance details
+        if log_env_summary:
+            logger.info(f"ep {ep}: {env_wrapper.summary}")
+
+        logger.info(
+            f"ep {ep} summary - "
+            f"running time: {time.time() - t0} "
+            f"env steps: {env_wrapper.step_index} "
+            f"experiences collected: {num_experiences_collected}"
+        )
+        if ep == eval_schedule[eval_point_index]:
+            # evaluation
+            eval_point_index += 1
+            logger.info("Evaluating...")
+            agent_wrapper.exploit()
+            eval_env_wrapper.reset()
+            eval_env_wrapper.start()  # get initial state
+            while eval_env_wrapper.state:
+                action = agent_wrapper.choose_action(eval_env_wrapper.state)
+                eval_env_wrapper.step(action)
+
+            # performance details
+            logger.info(f"Evaluation result: {eval_env_wrapper.summary}")
```

Comment:
how could the `proxy` of the `actor` be closed?

Reply:
The actor proxy needs to be closed when the actor has finished its loop. Let me add it.